src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,208 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public class ServerSentEventsMessageParser
+    {
+        const byte ByteCR = (byte)'\r';
+        const byte ByteLF = (byte)'\n';
+        const byte ByteSpace = (byte)' ';
+
+        private InternalParseState _internalParserState = InternalParseState.ReadMessageType;
+        private IList<byte[]> _data = new List<byte[]>();
+
+        public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            var messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    if(end != buffer.End)
+                    {
+                        return ParseResult.Incomplete;
+                    }
+                    throw new FormatException("Expected a '\r\n' line ending");
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
```

Comment:
I assume `lineEnd` was positioned at the `\n`, so this is advancing past it? If so, throw a comment down here to clarify that :)